### DISCUSSION TITLE
Apps that change their icons when they are launched.

### DISCUSSION THREAD

**Discussion by noideastowork** [elrumo/macOS_Big_Sur_icons_replacements#1500] [Help with the website]
Recently, I have come across a lot of programs that change the icon when they are running. 

For some, icon for activa state can be changed in the settings (nekoray, spark), but for others it is very difficult to find solutuon. For example, I still do not know how to change peazip icon. 

How can I find these "hidden active icons"? Where are they usually located?
How should I change icons in such a way that the new icon remains even after opening the app?

<img width="209" alt="image" src="https://github.com/user-attachments/assets/25860025-1695-4508-975a-3c97fabfa983">
<img width="224" alt="image" src="https://github.com/user-attachments/assets/c43e7344-6f2e-4ae2-a2d8-ffd34335b113">

**Comment by vikasbhatia** [comment]
I have a solution, but before you read further, just know that your mileage may vary.

You need to browse inside the .app package and find the .icns file inside that it's taking when the app is launched, and replace that.

I first noticed this behaviour with VLC media player, and found it very annoying as well, so I dug around in the folder structure.

In VLC's case, this was located in `VLC.app > Show Package Contents > Contents > Resources > VLC.icns` but this may be different based on the app you're using. If you don't see it in the Resources folder, you may have to try searching or poking around a little.

Then, overwrite that .icns with your custom one. I made a quick screen recording showing how this works:

https://github.com/user-attachments/assets/a2c2f8af-1479-4c3a-9766-8b93948ff941

**Disclaimer:**

- Icons replaced in this way (and not via an app like Pictogram) may get overridden with updates, so hang on to your custom .icns file for when you'll need to repeat this process.
- **__Always have a backup before trying this__** because you are technically modifying an app, which could potentially break its code signature and lead to unintended consequences (such as not recognising a license). In my case, you saw that VLC didn't give a shit, but some apps (from the App Store, for example) may complain that they were modified and are therefore unsafe, and refuse to launch.
- If that does happen, and you'd still like to replace their icon, you could look into force signing the app yourself to make those errors go away. It won't _harm_ your app per se, but you're basically replacing the signature and telling it that it's "whole" again. To do this, open a Terminal and run `sudo codesign -vvv --deep --strict App.app` where `App.app` is the path to the app you want to force sign. Easiest way to do it is to type the command up until `strict ` and then drag and drop the app onto the terminal window and it'll fill in the path.

Hope this helps!

**Comment by noideastowork** [comment]
Thank you so much for your detailed and great reply. Also, I am super impressed by your video and the effects on it! 

Unfortunately, I knew about Show Package Contents > Contents and I searched there. But none of the apps I had dealt with had this "active" icon inside any folder in the Contents.

So, this is exactly why I asked this question. I thought that most apps put the "active" icon somewhere outside. But now I have another hypothesis. Maybe all I have problems with are not truly macOS apps.

For example, I am currently looking for an icon in Peazip (I changed any icon or image I found in this folder - no result). Maybe it is because it is not a truly native app or something like that (Rosetta?).

Or maybe there's another case. I can't find the icon for the menu bar in app Karing. Perhaps it was developed for the first place on the iPad (and that is the reason).

So, in case you have any ideas, I would be very grateful. And again - thank you @vikasbhatia

**Comment by vikasbhatia** [comment]
Thanks! I used [Screen Studio](https://www.screen.studio/); it does all the fancy stuff automatically :)

Out of curiosity, I downloaded PeaZip to give it a go myself. It can't be that hard, right? (Famous last words).
Sorry about the long read below but I want you to know that I gave it a try, I really did 🥲

FYI I'm on an Intel mac and it sounds like you may be on Apple Silicon(?) and we're seeing the same thing, so I don't think it's to do with Rosetta. (Side note: why distribute an app with a proper icon that has rounded corners and looks totally fine inside the Applications folder, but then change the icon on run? Like, why?! You already made the good one! Sheesh.)

1. First I looked around for an .icns. Didn't find one. Okay, no problem.

2. Next: looked around for any PNGs or icon files, folder by folder.
In `/Applications/peazip.app/Contents/Resources/share/batch/freedesktop_integration` I found a square `peazip.png` there -- nice! Let's try changing that. So I converted the PeaZip icon into a PNG and replaced it. Relaunched PeaZip - no go.

3. I looked at a readme in the same folder in case it had any helpful info. It asks you to move some icons to a particular folder, which I tried, but it didn't work. It was worth a shot; this seemed relevant to Linux desktop environments anyway and not macOS.

![image](https://github.com/user-attachments/assets/1346e625-fce0-417a-a2d3-2f5dd6ecc899)

4. In `/Applications/peazip.app/Contents/Resources/share/icons` I found some PNGs and an .ico file that looked promising.

![image](https://github.com/user-attachments/assets/0c555f02-5230-492d-b027-60f4fe2c85fc)

So I made a rounded version of the .ico, replaced them all, opened PeaZip again, and behind door number 1...nope! Still didn't work.

That was when I realised that the files I replaced still had a little rounding in the corners and the ugly square dock icon had completely sharp edges.

![peazip pls](https://github.com/user-attachments/assets/db7c0744-8283-49ee-b679-02dfba4d6db0)

So it's obviously getting that sharp-edged icon from somewhere else, and those files I replaced were serving no apparent purpose and were planted there as a red herring to taunt us (obviously).

5. I googled "peazip mac square dock icon reddit" because there's probably someone there who's managed to do it, and I found [this thread](https://redd.it/1eo8n7i) (which I think may be you? 🙈 haha) and there were sadly no answers yet. I looked at the subreddit size and...there are only 389 people on that subreddit. Admittedly 389 very passionate people, to have cared about PeaZip enough to have subscribed to the subreddit lol, but it figures that there aren't any answers yet.

6. So I'd have to find where that icon's coming from, as there was nothing else inside the .app. PeaZip is open-source, so I went to [the source](https://github.com/peazip/PeaZip/tree/sources/peazip-sources) and I found it (!) in `/PeaZip-sources/PeaZip.ShellEx/PeaZip.Package/Assets/`. There it is, alongside its rounded counterparts. It _sounds_ like that's for the shell extension/right click menu and may not be relevant to the dock icon, but my strategy was to just carpet-bomb replace any ugly sharp edges icon I find and see what sticks lol.

![moby dick](https://github.com/user-attachments/assets/c92f4005-cfef-4023-a082-baa1d0bad72e)

Great, how do I replace it? F*** it, we'll fix it and build our own copy of the app from source! I looked at the language to see what the requirements are to build it, annnnnnd the app is coded in, of all things, **_Pascal._**

That was where I decided to call it a night 😆 I'm sorry, but it looks like that square icon may be baked into the app somehow.

PS. I considered opening an issue on GitHub to tell them they should fix this, but for some reason I couldn't even find Issues on their repo. Maybe they have it disabled? All they have is a Pull Requests tab; even https://github.com/peazip/PeaZip/issues redirects there. Which means your options are:

a) [Email the developer](giorgio.tani.software@gmail.com).
b) Code the change yourself an open a PR on GitHub.
c) Just use [The Unarchiver](https://theunarchiver.com/) or [Keka](https://www.keka.io/en/) and change the icon to PeaZip out of spite.

I'm leaning towards option C.

**Comment by noideastowork** [comment]
I appreciate all your efforts you have made to help me! Really! I am super impressed as you basically reproduced all the actions I did and even some more! Wow!

1) Yeah, that is my thread, haha. In all cases, I try to write about staff I have problems with to support groups, GitHub issues, Reddit. I wrote about all the apps I mentioned before (Nekoray, Spark, Karing and etc.). Sometimes it helps, sometimes not.

2) In fact, they have PeaZip Tickets on [sourceforge](https://sourceforge.net/p/peazip/tickets/). Maybe it's worth a try. But I couldn't find a button to create ticket (probably because I didn't log in).

3) Yes, it can be compiled, somehow. But honestly, I'm not even a developer, so I'm not sure how to do it. I am kinda stupid for it. That means I need to put a lot of effort into finding a way to do it.

> As alternative, PeaZip can be [compiled and built from sources](https://peazip.github.io/peazip-sources.html) (Open Source, LGPLv3 license), with Lazarus/FPC IDE 2.2RC or more recent, as native application for Darwin / macOS, both for Intel x86_64 CPU (OSX up to macOS 10.15 Catalina), and for Apple Silicon M1 aarch64 ARM SoC (for macOS 11 Big Sur, Monterey and more recent systems), targeting in both cases Cocoa widgetset.
> Legacy Carbon widgetset is still supported by Lazarus, but untested regarding compiling and building PeaZip for legacy Apple systems.
> 
> For best results compiling project_peach.lpi and project_pea.lpi with Lazarus/FPC targeting Macintosh machines, open main menu, Project > Project Options, Configuration and Target, and set Darwin as destination OS, and aarch64 as destination CPU family if compiling for machines with M1, M2 Apple Silicon chip family, or x86_64 if compiling for Intel-based Apple machines.

4) You're right. In the case of PeaZip, the best solution is to send an email (or post a ticket on SourceForge) and switch to another app. There are plenty of archiving software options available. There are The Unarchiver, Keka, MacZip, or other app.

As for Karing, I think I found the icon for the menu bar (called "tray"). I'll try to change it and see how it works.

P.S. Hm, Screen Studio - I am going to remember this. 

P.P.S. And thanks again! I have no words to describe all the respect and gratitude I feel for you.
